Sources/System/IORing.swift

```diff
@@ -0,0 +1,657 @@
+            )
+        } else {
+            return ret
+        }
+    }
+}
+
+internal func _submitRequests(ring: borrowing SQRing, ringDescriptor: Int32) throws {
+    let flushedEvents = _flushQueue(ring: ring)
+    _ = try _enter(
+        ringDescriptor: ringDescriptor, numEvents: flushedEvents, minCompletions: 0, flags: 0)
+}
+
+internal func _getUnconsumedSubmissionCount(ring: borrowing SQRing) -> UInt32 {
+    return ring.userTail - ring.kernelHead.pointee.load(ordering: .acquiring)
+}
+
+internal func _getUnconsumedCompletionCount(ring: borrowing CQRing) -> UInt32 {
+    return ring.kernelTail.pointee.load(ordering: .acquiring)
+        - ring.kernelHead.pointee.load(ordering: .acquiring)
+}
+
+//TODO: pretty sure this is supposed to do more than it does
+internal func _flushQueue(ring: borrowing SQRing) -> UInt32 {
+    ring.kernelTail.pointee.store(
+        ring.userTail, ordering: .releasing
+    )
+    return _getUnconsumedSubmissionCount(ring: ring)
+}
+
+@inline(__always)
+internal func _getSubmissionEntry(
+    ring: inout SQRing, submissionQueueEntries: UnsafeMutableBufferPointer<io_uring_sqe>
+) -> UnsafeMutablePointer<
+    io_uring_sqe
+>? {
+    let next = ring.userTail &+ 1  //this is expected to wrap
+
+    // FEAT: smp load when SQPOLL in use (not in MVP)
+    let kernelHead = ring.kernelHead.pointee.load(ordering: .acquiring)
+
+    // FEAT: 128-bit event support (not in MVP)
+    if next - kernelHead <= ring.array.count {
+        // let sqe =  &sq->sqes[(sq->sqe_tail & sq->ring_mask) << shift];
+        let sqeIndex = Int(
+            ring.userTail & ring.ringMask
+        )
+
+        let sqe = submissionQueueEntries
+            .baseAddress.unsafelyUnwrapped
+            .advanced(by: sqeIndex)
+
+        ring.userTail = next
+        return sqe
+    }
+    return nil
+}
+
+public struct IORing: ~Copyable {
+    let ringFlags: UInt32
+    let ringDescriptor: Int32
+
+    @usableFromInline var submissionRing: SQRing
+    // FEAT: set this eventually
+    let submissionPolling: Bool = false
+
+    let completionRing: CQRing
+
+    let submissionQueueEntries: UnsafeMutableBufferPointer<io_uring_sqe>
+
+    // kept around for unmap / cleanup
+    let ringSize: Int
+    let ringPtr: UnsafeMutableRawPointer
+
+    var _registeredFiles: [UInt32]?
+    var _registeredBuffers: [iovec]?
+
+    public init(queueDepth: UInt32) throws {
+        var params = io_uring_params()
+
+        ringDescriptor = withUnsafeMutablePointer(to: &params) {
+            return io_uring_setup(queueDepth, $0)
+        }
+
+        if params.features & IORING_FEAT_SINGLE_MMAP == 0
+            || params.features & IORING_FEAT_NODROP == 0
+        {
+            close(ringDescriptor)
+            // TODO: error handling
+            throw IORingError.missingRequiredFeatures
+        }
+
+        if ringDescriptor < 0 {
+            // TODO: error handling
+        }
```

Comment:
should this move to line 221 before:

```
if params.features & IORING_FEAT_SINGLE_MMAP == 0
             || params.features & IORING_FEAT_NODROP == 0
         {
             close(ringDescriptor)
```

Reply:
I want to redo all the feature detection stuff but I'll take a look at this when I do